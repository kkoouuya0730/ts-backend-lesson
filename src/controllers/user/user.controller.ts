import * as userService from "../../services/user/user.service";
import { Request, Response, NextFunction } from "express";
import { AppError } from "../../errors/AppError";
import { INVALID_USER_ID, INVALID_USER_INPUT, USER_NOT_FOUND } from "../../constants";
import { userIdSchema, userInputSchema } from "../../validation";

export const getUsersHandler = async (_req: Request, res: Response, next: NextFunction) => {
  try {
    const users = await userService.getUsers();
    res.json(users);
  } catch (error) {
    next(error);
  }
};

export const getUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  console.log(parsedUserIdResult);
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    const user = await userService.getUser(userId);
    if (!user) {
      throw new AppError(404, USER_NOT_FOUND);
    }
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const updateUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });

  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  const parsedInput = userInputSchema.safeParse(req.body);
  if (!parsedInput.success) {
    throw new AppError(400, INVALID_USER_INPUT, parsedInput.error.issues);
  }

  try {
    const user = await userService.updateUser(userId, parsedInput.data.name);
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const deleteUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    await userService.deleteUser(userId);
    res.json({ message: "User deleted successfully" });
  } catch (error) {
    next(error);
  }
};
