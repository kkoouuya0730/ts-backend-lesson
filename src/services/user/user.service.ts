import prisma from "../../models/prisma";
import { UserInput } from "../../validation";

export const getUsers = async () => {
  const users = await prisma.user.findMany();
  return users.map(({ passwordHash, ...publicUser }) => publicUser);
};

export const getUser = async (id: number) => {
  const user = await prisma.user.findUnique({
    where: {
      id,
    },
  });
  if (!user) return null;
  const { passwordHash, ...publicUser } = user;
  return publicUser;
};

export const createUser = async (data: UserInput) => {
  return await prisma.user.create({
    data,
  });
};

export const updateUser = async (id: number, name: string) => {
  const updatedUser = await prisma.user.update({
    where: {
      id,
    },
    data: {
      name,
    },
  });
  const { passwordHash, ...publicUser } = updatedUser;
  return publicUser;
};

export const deleteUser = async (id: number) => {
  await prisma.user.delete({
    where: {
      id,
    },
  });
};
