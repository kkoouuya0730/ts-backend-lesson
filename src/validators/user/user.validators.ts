import { ZodSafeParseResult } from "zod";
import { userIdSchema, UserInput, userInputSchema } from "../../validation";

/**
 * 例外を投げる関数を限定するという意図のもと、
 * parse系関数では例外を投げず、呼び出しもとで例外処理を行う
 */

export const parseUserId = (
  id: string | undefined
): ZodSafeParseResult<{
  id: number;
}> => {
  const parsedUserIdResult = userIdSchema.safeParse({ id });
  return parsedUserIdResult;
};

export const parseUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};

export const parseCreateUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};
