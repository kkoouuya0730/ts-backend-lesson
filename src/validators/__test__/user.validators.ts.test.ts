import { parseUserId, parseUserInput } from "../user.validators";

describe("parseUserId", () => {
  test("有効な数値文字列を受け取ったら成功する", () => {
    const result = parseUserId("123");
    expect(result.success).toBe(true);
    expect(result.data).toEqual({ id: 123 });
    expect(result.data).toMatchSnapshot();
  });

  test.each(["0", "-1", "abc", undefined, ""])("無効な入力(%s)は失敗する", (input) => {
    const result = parseUserId(input as any);
    expect(result.success).toBe(false);
    expect(result.data).toMatchSnapshot();
  });
});

describe("parseUserInput", () => {
  test("有効な文字列を受け取ったら成功する", () => {
    const result = parseUserInput({ name: "aaa" });
    expect(result.success).toBe(true);
    expect(result.data).toEqual({ name: "aaa" });
    expect(result.data).toMatchSnapshot();
  });

  test("有効な文字数(50文字以下)の文字列を受け取ったら成功する", () => {
    const result = parseUserInput({
      name: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    });
    expect(result.success).toBe(true);
    expect(result.data).toEqual({
      name: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    });
    expect(result.data).toMatchSnapshot();
  });

  test.each([
    { input: { name: "" }, desc: "空文字列" },
    { input: { name: 123 as any }, desc: "数値" },
    { input: undefined, desc: "undefined" },
  ])("無効な入力(%s)は失敗する", ({ input }) => {
    const result = parseUserInput(input as any);
    expect(result.success).toBe(false);
    expect(result.data).toMatchSnapshot();
  });
});
