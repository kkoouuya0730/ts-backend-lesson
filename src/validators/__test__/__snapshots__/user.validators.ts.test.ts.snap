// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`parseUserId 有効な数値文字列を受け取ったら成功する 1`] = `
{
  "id": 123,
}
`;

exports[`parseUserId 無効な入力()は失敗する 1`] = `undefined`;

exports[`parseUserId 無効な入力(-1)は失敗する 1`] = `undefined`;

exports[`parseUserId 無効な入力(0)は失敗する 1`] = `undefined`;

exports[`parseUserId 無効な入力(abc)は失敗する 1`] = `undefined`;

exports[`parseUserId 無効な入力(undefined)は失敗する 1`] = `undefined`;

exports[`parseUserInput 有効な文字列を受け取ったら成功する 1`] = `
{
  "name": "aaa",
}
`;

exports[`parseUserInput 有効な文字数(50文字以下)の文字列を受け取ったら成功する 1`] = `
{
  "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
}
`;

exports[`parseUserInput 無効な入力({ input: [Object], desc: '数値' })は失敗する 1`] = `undefined`;

exports[`parseUserInput 無効な入力({ input: [Object], desc: '空文字列' })は失敗する 1`] = `undefined`;

exports[`parseUserInput 無効な入力({ input: undefined, desc: 'undefined' })は失敗する 1`] = `undefined`;
